core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -318,6 +323,16 @@ private Object processWithExceptionHandling(Expression expression, Object contex
             }
         }
 
+        private boolean canUseCache(Object context)
+        {
+            // We are using WeakHashMap as cache, that's why we can't depend on an intermediate object
+            // that consists of expression as well as context as a key. This is because intermediate object could
+            // get GCed even when expression still exists which will cause cache miss. So, to solve this problem we
```

Comment:
I mean if we create an intermediate object which contains context and expression as a key for WeakHashMap to avoid this check `context instanceof NoOpSymbolResolver`. In that case, it can get GCed right?

For ex: https://github.com/trinodb/trino/pull/12016#discussion_r853018269